File: solutions/problem_1769_minimum_number_of_operations_to_move_all_balls_to_each_box/solution.cpp
/*
 * Problem 1769: Minimum Number of Operations to Move All Balls to Each Box
 * ======================================================================
 * Difficulty: Medium
 * Tags: Array, String, Prefix Sum
 * Pattern: Prefix Sum
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> minOperations(string& boxes) {
        // Prefix sum approach - O(n) time, O(n) space
        unordered_map<int, int> prefix;
        prefix[0] = -1;
        int curr_sum = 0, result = 0;
        int target = boxes;
        for (int i = 0; i < (int)boxes.size(); i++) {
            curr_sum += boxes[i];
            if (prefix.count(curr_sum - target)) {
                result = max(result, i - prefix[curr_sum - target]);
            }
            if (!prefix.count(curr_sum)) {
                prefix[curr_sum] = i;
            }
        }
        return result;
    }
};

File: solutions/problem_1769_minimum_number_of_operations_to_move_all_balls_to_each_box/solution.py
"""
Problem 1769: Minimum Number of Operations to Move All Balls to Each Box
======================================================================
Difficulty: Medium
Tags: Array, String, Prefix Sum
Pattern: Prefix Sum

Time Complexity:  O(n)
Space Complexity: O(n)
"""

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        # Prefix sum approach - O(n) time, O(n) space
        prefix = {0: -1}
        curr_sum = 0
        result = 0
        target = boxes if isinstance(boxes, int) else 0
        for i, val in enumerate(boxes):
            curr_sum += val
            if curr_sum - target in prefix:
                result = max(result, i - prefix[curr_sum - target])
            if curr_sum not in prefix:
                prefix[curr_sum] = i
        return result
